### ISSUE TITLE
Possibly bug, TS wrongly complains that generic property is null

### ISSUE THREAD

**Issue by alexeypetrushin** [Microsoft/TypeScript#30205]
TS v 3.3.3, **enable strictNullChecks**

**Code**

```ts
// Errors is a collection of form errors
//
// const errors = { 
//   title: ["can't be blank", "invalid characters"],
//   text:  ["too long"],
// }
//
// It could be rendered as
// 
// <FormInput name="title" errors={errors}/>


// Requiring that Errors had field of type T
type Errors<T extends string> = { [k in T]?: string[] }

export function FormInput<T extends string>(
  { name, errors }: { name: T, errors: Errors<T> }
) {
  // TS complains that there's no length
  const hasErrors = (errors[name] || []).length > 0
}
```

**Expected behavior:**

Should correctly infer that `(errors[name] || [])` is an Array.

**Actual behavior:**

Wrongly complains about `null`

**Playground Link:** 

[Play](https://www.typescriptlang.org/play/#src=%2F%2F%20Errors%20is%20a%20collection%20of%20form%20errors%0A%2F%2F%0A%2F%2F%20const%20errors%20%3D%20%7B%20%0A%2F%2F%20%20%20title%3A%20%5B%22can't%20be%20blank%22%2C%20%22invalid%20characters%22%5D%2C%0A%2F%2F%20%20%20text%3A%20%20%5B%22too%20long%22%5D%2C%0A%2F%2F%20%7D%0A%2F%2F%0A%2F%2F%20It%20could%20be%20rendered%20as%0A%2F%2F%20%0A%2F%2F%20%3CFormInput%20name%3D%22title%22%20errors%3D%7Berrors%7D%2F%3E%0A%0A%0A%2F%2F%20Requiring%20that%20Errors%20had%20field%20of%20type%20T%0Atype%20Errors%3CT%20extends%20string%3E%20%3D%20%7B%20%5Bk%20in%20T%5D%3F%3A%20string%5B%5D%20%7D%0A%0Aexport%20function%20FormInput%3CT%20extends%20string%3E(%0A%20%20%7B%20name%2C%20errors%20%7D%3A%20%7B%20name%3A%20T%2C%20errors%3A%20Errors%3CT%3E%20%7D%0A)%20%7B%0A%20%20const%20hasErrors%20%3D%20(errors%5Bname%5D%20%7C%7C%20%5B%5D).length%20%3E%200%0A%7D)

**Comment by RyanCavanaugh** [comment]
Duplicate #22137 - we don't have type system machinery in place yet to narrow away `undefined`/`null` from types originating in a generic type.
